### DISCUSSION TITLE
MicroPython presentation at Teardown2024!

### DISCUSSION THREAD

**Discussion by bikeNomad** [micropython/micropython#15243] [General]
I’m excited to share that I (@bikeNomad, a/k/a @ned-pcs) will be presenting a 50 minute long talk titled [_Prototype to product with MicroPython: faster and happier embedded systems development_ ](https://www.crowdsupply.com/teardown/portland-2024/long-talk/prototype-to-product-with-micropython-faster-and-happier-embedded-systems-development?public=1)at #Teardown2024, @crowd_supply annual hardware hacking conference, June 21-23rd in Portland, Oregon. (The actual time and day of my talk is still TBD)

I'll be showing how to take your project from a simple concept to a sophisticated, streamlined product using MicroPython.

Among the topics I'll be covering are:

  - Why use MicroPython?
  - Useful third party libraries that can save lots of time
  - Custom board definitions and partition layouts
  - Pre-compiling and freezing code
  - Saving memory
  - Optimization using the viper and native code generators
  - Over-the-air updating
  - Packaging third party libraries in your project
  - Custom C/C++ modules
  - Interrupts, threading, and asyncio
  - Debugging and logging strategies

Get your tickets now! https://www.crowdsupply.com/teardown/portland-2024

**Comment by bikeNomad** [comment]
I just got the schedule slot: it will be at 11:30 on Saturday 22 June. Hope to see you all there!

**Comment by mattytrentini** [comment]
Wish I could be there, look forward to seeing the recording! Hope it all goes smoothly. 😄 

If you'd like someone to review your presentation please just reach out...

**Reply by bikeNomad** [reply]
Thanks! I'm still working on the presentation but would appreciate your comments. I'll reach out to you in a few days.

**Comment by jonnor** [comment]
Sounds like a great talk. I hope it will be available online later, for those that are not able to make it to Portland?

**Reply by bikeNomad** [reply]
They will be making the videos available online; I don't know when or where yet. Will keep you posted!

**Comment by ned-pcs** [comment]
[The video is available on YouTube](https://www.youtube.com/watch?v=Z9PnyaDmQjM&ab_channel=CrowdSupply).

**Reply by PolarGoose** [reply]
[slides](https://www.crowdsupply.com/files/6dae/5ed80797-e301-40d6-ac73-71f549146dae/teardown-2024-ned-konz-micropython.pdf)

Amazing talk. I have learned a lot.
The only thing is that I think the `Debugging and logging strategies` topic wasn't mentioned. That is what I'm struggling with. I want the same luxury as on the Desktop, where I can have a detailed debug log available. Something similar to Linux syslog would be great. However, I haven't found anything like that for Micropython. Even a simple rotating log file is not for free. I had to make it myself: [RotatingFileHandler](https://github.com/PolarGoose/MicroPython-Logging-RotatingFileHandler). But then I can't access the logs while my program is running because while I'm reading a log file, my program might want to do the file rotation, and it will fail because the reading process uses the file.
Do you have any hints on how to implement debug logging?

**Reply by ned-pcs** [reply]
I use a patched version of the CPython `logging` module from `micropython-lib` (see https://github.com/micropython/micropython-lib/pull/806) in my projects. Do you have to read the log files while running? You can add a `StreamHandler` to your logger to provide for interactive reading (connect it to a circular buffer). While I don't do rotation while logging, this can be done by removing the `FileHandler` connected to your old log file and re-adding another `FileHandler` connected to your new log file. I have used a simple `StreamHandler` class (`asyncmd.aioctl.log`) from @Carglglz `asyncmd` library (https://github.com/Carglglz/asyncmd); this provides a way to view the last 2K of your log.

**Reply by PolarGoose** [reply]
@ned-pcs,

Great, thank you for the suggestions. A circular buffer would work very well. I looked at the [aiolog](https://github.com/Carglglz/asyncmd/blob/main/aiolog.py), and it looks promising.

> Do you have to read the log files while running?

It is not mandatory. Therefore, I thought about making a script that will stop the program by renaming `main.py` to `_main.py` and restarting the board then copy the log files, and revert the board back to its operational state. In this way I can easily get megabytes of debug logging, which is big enough to log everything I need, and in case my board stops working, I can just get the logs and see what is wrong.

**Reply by Carglglz** [reply]
@ned-pcs 
> this provides a way to view the last 2K of your log.

It can be increased  if needed e.g. `AioStream(4000)` 

@PolarGoose I use both `aiolog` and [rsysloghandler](https://github.com/Carglglz/asyncmd/tree/main/tools#rsysloghandler) as stream handlers for remote debugging (in case your board has network capability) 👍🏼
